### ISSUE TITLE
HTTPS Connection hangs when using Connect-Proxy with Android Emulator

### ISSUE THREAD

**Issue by NghiaTranUIT** [apple/swift-nio-examples#88]
Hi, this week, I've encountered some strange issues when trying proxy the HTTPS connection from Android Emulator. I tried, but I haven't found an answer yet. I hope by opening this ticket here, someone can help me. 

Thanks for your hard work on the SwiftNIO project 🙇 

### Expected behavior
- The HTTPS connection from Android Emulator should works fine.

### Actual behavior
- All connections, which are called by Android Emulator, hang forever until the timeout.
- `8.8.8.8:853` is hang

### Steps to reproduce
1. Use the Connect-Proxy example
2. Listen on 0.0.0.0 at port 8080
3. Add some useful log into GlueHandler

![CleanShot 2023-08-19 at 20 53 06@2x](https://github.com/apple/swift-nio-examples/assets/5878421/9558855f-dabe-4664-bb9b-f6bf162e39c9)

4. Start the main.swift 
5. Verify the connect-proxy is working fine ✅  : `curl -v "https://httpbin.proxyman.app/get?id=123" --proxy localhost:8080`

![CleanShot 2023-08-19 at 20 54 27@2x](https://github.com/apple/swift-nio-examples/assets/5878421/053c9fde-aa2c-4d6e-aa46-8a49c77e852f)

7. Verify the console log: the GlueHandler is added and `channedRead` is called ✅ 
```
⚡️ [GlueHandler] handlerAdded
🔍 [GlueHandler] channelRead, data = \(data)
```

8. Open Android Emulator -> Set Setting Proxy to current IP

![CleanShot 2023-08-19 at 20 47 20@2x](https://github.com/apple/swift-nio-examples/assets/5878421/bcb38188-5961-48b5-8994-f226a4d9717d)

10. Open Google Chrome -> Visit any website
11. Observe that there are a lot of waiting connections: The GlueHandler is added, but there is no readable data -> The socket hangs forever => ❌ 🐛 Bug here 

![CleanShot 2023-08-19 at 20 51 23@2x](https://github.com/apple/swift-nio-examples/assets/5878421/070d7ace-7a10-44ec-82a2-1d7547157ef5)

### 📝 My investigation
1. After setting the HTTP Proxy from Android Emulator, it seems the Android is using DNS over HTTPS (Private DNS is enabled)
2. When open Google Chrome, visit any website -> Android makes a CONNECT request, without the Host Header.
```
CONNECT 8.8.8.8:853 HTTP/1.1
```
3. `connect-proxy` returns `HTTP/1.1 200 OK`
4. but for some reason, the client doesn't send any data. The socket just hangs.

I guess the connect-proxy doesn't support TCP yet, but if you don't mind, can you point out how to support it? I tried to add a new TCPHandler to read the data, but there is no readable data is called.

### If possible, minimal yet complete reproducer code (or URL to code)
Video: https://drive.google.com/file/d/1ojOtRa-15N_pPklrApIFAAhX3xYfeQ2i/view?usp=sharing 

### version/commit hashes from all involved dependencies

The latest version of connect-proxy example.

**Comment by NghiaTranUIT** [comment]
So, I investigate further by writing a dead simple Proxy by NodeJS.

It turns out, `connect-proxy` returns the `Content-Length: 0` in the `HTTP/1.1 200 OK` message ([code](https://github.com/apple/swift-nio-examples/blob/caaded623747da68cc3dd34238b657593f94cd16/connect-proxy/Sources/ConnectProxy/ConnectHandler.swift#L207)), so Android Emulator doesn't work 🤔 

cURL or modern network client (URLSession, Web browsers) works fine, but not Android Emulator.

For example: cURL ignores it.
```
< HTTP/1.1 200 Connection Established
< Content-Length: 0
* Ignoring Content-Length in CONNECT 200 response
```

---------------------

Question:
1. Is it possible to completely remove the `Content-Length: 0` or `Transfer-Encoding: chunked` ? I just want to return the simple message `HTTP/1.1 200 OK` from the CONNECT request.

**Comment by Lukasa** [comment]
Yeah, good catch. Right now NIO doesn't support doing that, but it should. It's a somewhat complex patch to write though. Let me see if I can throw it together.

**Comment by NghiaTranUIT** [comment]
Thanks. I finally write a patched version of HTTPResponseEncoder. It removed the `Content-Length`, and `Transfer-Encoding` logic. In the end, It fixes the bug.

Not sure if there are any unseen problems? 🤔

**Comment by Lukasa** [comment]
The problems are in the general use-case: it isn't safe for us to unconditionally omit them, because users rely on our behaviour of inserting them. The trick is that we only want to skip that logic for responses to CONNECT requests. Fortunately, I believe we have the hooks we need to do this.

**Comment by Lukasa** [comment]
Oof, actually it's quite hard to do it nicely. I think we need to add configuration to the encoders to enable you to opt-out of them modifying the framing headers.

**Comment by NghiaTranUIT** [comment]
Thanks. I will keep on eyes on it, and report if it has any problems on production build 👍

**Comment by Lukasa** [comment]
See apple/swift-nio#2508.

**Comment by Lukasa** [comment]
I just merged that patch to `main`, want to try to use it?

**Comment by NghiaTranUIT** [comment]
Thanks so much for your work @Lukasa I'm going to try it now 👍

**Comment by NghiaTranUIT** [comment]
It works as intended. I'm happy to close this ticket 🙇
